### PR TITLE
Add caching for exclusion list

### PR DESCRIPTION
**Pull request by SlayterDev**
**Reviewer:** @sammacbeth @jonathanKingston @kdzwinel 

## Description:
Cookie exclusion list was not being cached. If the list endpoint returned a 304 there would be no response and the list would fail to load. This PR adds caching based on the UserAgent list implementation.

## Steps to test this PR:
1. Load extension
2. Visit site with excluded cookie domain (e.g. google 3rd party login)
3. Login should work
4. Reload extension (to force reloading exclusion list)
5. Repeat steps 2-3
**Note:** Since this relies on getting a 304 response you may need to reload the extension in quick succession

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

**Review by sammacbeth — COMMENTED** [review]
Why do we not just store this data in `chrome.storage.local`? This is a simpler API and less error prone than indexeddb IMO.

**Comment by SlayterDev** [comment]
@sammacbeth ready for review again.

File: shared/js/background/chrome-events.es6.js [hunk 1]
```diff
@@ -389,12 +389,13 @@ chrome.runtime.onMessage.addListener((req, sender, res) => {
             const parsed = tldts.parse(tabUrl)
             action.tabRegisteredDomain = parsed.domain === null ? parsed.hostname : parsed.domain
 
+            action.isThirdParty = !utils.isFirstParty(sender.url, sender.tab.url)
+            action.shouldBlock = !cookieConfig.isExcluded(sender.url)
+
             if (req.documentUrl && trackerutils.isTracker(req.documentUrl) && sender.frameId !== 0) {
-                action.isTrackerFrame = true
+                action.isTrackerFrame = true && action.shouldBlock
```
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:396]: This change seems unrelated to the list caching. Did you mean to include it?

**Review by sammacbeth — COMMENTED** [review]
Storage changes look good - I'm just a bit confused about the other changes in chrome events.

File: shared/js/background/chrome-events.es6.js (hunk 1 above)
Comment (SlayterDev) [reply shared/js/background/chrome-events.es6.js:396]: I did. The way the content script was set up was ignoring exclusions. This will check if the url is on the exclusion list then apply it to `isTrackerFrame` accordingly. Google login for slack was broken before this change.

Comment (sammacbeth) [reply shared/js/background/chrome-events.es6.js:396]: Ok, rather than changing the semantics of `isTrackerFrame`, can you maybe change the check here to not inject if `shouldBlock` is false? https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/content-scripts/block-cookie.js#L132

File: shared/js/content-scripts/block-cookie.js [hunk 2]
```diff
@@ -129,8 +129,8 @@
         checkThirdParty: true,
         documentUrl: window.location.href
     }, function (action) {
-        if (!action) {
-            // response is undefined if the background has not yet registered the message listener
+        if (!action || !action.shouldBlock || !action.isThirdParty) {
+            // action is undefined if the background has not yet registered the message listener
             return
         }
         if (window.top !== window && action.isTrackerFrame) {
```
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:136]:
```suggestion
        if (window.top !== window && action.isTrackerFrame && action.shouldBlock) {
```

File: shared/js/content-scripts/block-cookie.js [hunk 3]
```diff
@@ -129,8 +129,8 @@
         checkThirdParty: true,
         documentUrl: window.location.href
     }, function (action) {
-        if (!action) {
-            // response is undefined if the background has not yet registered the message listener
+        if (!action || !action.shouldBlock || !action.isThirdParty) {
```
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:132]: This should be reverted as it will break first party cookie code.

**Review by sammacbeth — CHANGES_REQUESTED** [review]

**Review by sammacbeth — APPROVED** [review]
Thanks for addressing the comments!
